shared/src/commonMain/kotlin/io/customer/shared/tracking/queue/QueueDispatcher.kt

```diff
@@ -0,0 +1,154 @@
+
+    private suspend fun acquireLock() {
+        logger.debug("Acquiring lock for queue")
+        kotlin.runCatching {
+            mutex.lock(owner = this)
+        }.onFailure { ex ->
+            logger.error("Mutex locking failed, reason: $ex")
+        }
+    }
+
+    private fun releaseLock() {
+        logger.debug("Releasing lock for queue")
+        kotlin.runCatching {
+            mutex.unlock()
+        }.onFailure { ex ->
+            logger.error("Mutex unlocking failed, reason: $ex")
+        }
+    }
+
+    override fun onCoroutineFailed(exception: Throwable) {
+        releaseLock()
+    }
+
+    override suspend fun checkForPendingTasks(source: QueueTriggerSource) {
+        processBatchTasks(isRecursive = false) { getTasksToBatch() }
+    }
+
+    override suspend fun sendAllPending() {
+        processBatchTasks(isRecursive = true) {
+            trackingTaskQueryHelper.selectAllPendingTasks().getOrNull()
+        }
+    }
+
+    /**
+     * Since we have imposed a limit of maximum number of tasks, [isRecursive] allows the caller
+     * to send multiple batch events. If true, it will keep calling [getTasks] after every
+     * successful execution until an empty list of tasks is provided.
+     */
+    private fun processBatchTasks(
+        isRecursive: Boolean,
+        getTasks: suspend () -> List<TrackingTask>?,
+    ) = runOnBackground(
+        onFailure = { releaseLock() },
+    ) {
+        if (mutex.isLocked) return@runOnBackground
+
+        acquireLock()
+        kotlin.runCatching {
+            queueTimer.cancel()
+        }.onFailure { ex ->
+            logger.error("Cannot cancel scheduled timer error: ${ex.message}")
+        }
+        kotlin.runCatching {
+            do {
+                val pendingTasks = getTasks()
+                val isSuccess: Boolean = if (!pendingTasks.isNullOrEmpty()) {
+                    queueRunner.runQueueForTasks(pendingTasks).getOrNull() == true
+                } else false
+            } while (isRecursive && isSuccess)
+        }.onFailure { ex ->
+            logger.error("Validation for pending tasks failed with error: ${ex.message}")
+        }
+        kotlin.runCatching {
+            queueTimer.schedule(
+                force = true,
+                duration = TimeUnit.Seconds(backgroundQueueConfig.batchDelayMaxDelayInSeconds.toDouble()),
+            ) { runOnBackground { checkForPendingTasks(QueueTriggerSource.TIMER) } }
+        }.onFailure { ex ->
+            logger.error("Cannot schedule timer error: ${ex.message}")
+        }
+        releaseLock()
+    }
+
+    private fun shouldSendImmediately(task: TrackingTask, timestamp: Instant): Boolean {
+        val timeInQueue = timestamp.minus(task.createdAt).toLong(DurationUnit.SECONDS)
+        return shouldCountInBatchTrigger(task) && (
+                timeInQueue >= backgroundQueueConfig.batchDelayMaxDelayInSeconds
+                        || task.priority > Priority.DEFAULT)
```

Comment:
```suggestion
  if (task.priority > Priority.DEFAULT) return true 
  
  val timeInQueue = timestamp.minus(task.createdAt).toLong(DurationUnit.SECONDS)
  if (timeInQueue >= backgroundQueueConfig.batchDelayMaxDelayInSeconds) return true 
  
 return shouldCountInBatchTrigger(task)
```

I find guard statements would make this code easier to read/maintain.

Reply:
I agree, if we are okay with multiple returns then we can surely update this. I personally don't prefer multiple returns in larger methods. But here, the method is small and this can help readability.

Dropping your suggestion because this changes the initial logic i.e. if the task was failed previously for reason any other than connectivity and is fixable, we do not run the queue specially for them assuming it will not add any value. So `shouldCountInBatchTrigger` will run first to make sure the task is eligible to trigger the queue.

I will update this with guard statements to make more readable.